index.js

```diff
@@ -500,6 +509,38 @@ module.exports = class NoiseSecretStream extends Duplex {
     cb(null)
   }
 
+  async _boxMessage (buffer) {
+    if (!this._boxSecret) await new Promise(resolve => this.once('handshake', resolve))
+    const nonce = this._boxNonce
+    sodium.sodium_increment(nonce)
+    const envelope = b4a.allocUnsafe(nonce.length + buffer.length + sodium.crypto_secretbox_MACBYTES)
+    sodium.crypto_secretbox_easy(envelope.subarray(nonce.length), buffer, nonce, this._boxSecret)
+    envelope.set(this._boxNonce)
+    return envelope
+  }
+
+  async send (buffer) {
+    if (!this.rawStream?.send) return // udx-stream expected
+    const message = await this._boxMessage(buffer)
+    return this.rawStream.send(message)
+  }
+
+  trySend (buffer) {
+    if (!this.rawStream?.trySend) return // udx-stream expected
+    this._boxMessage(buffer)
+      .then(message => this.rawStream.send(message))
+  }
+
+  async _onmessage (buffer) {
+    if (!this._boxSecret) await new Promise(resolve => this.once('handshake', resolve))
```

Comment:
would just drop it if the handshake isnt done and have this be sync (its 1 RT in practice so requires intense reordering to not be the case and even then its udp so we can do stuff like that)

Reply:
Yes that simplifies things, i had a note somewhere that messages should be buffered. But as you say, it's UDP. Critical messages should go through write/data not send/message